src/Razor/test/Microsoft.VisualStudio.Editor.Razor.Test/Settings/WorkspaceEditorSettingsTest.cs

```diff
@@ -0,0 +1,113 @@
+        settingsManagerMock
+            .Setup(x => x.GetClientSettings())
+            .Returns(expectedSettings);
+
+        // Act
+        var editorSettings = new WorkspaceEditorSettings(settingsManagerMock.Object);
+
+        // Assert
+        Assert.Equal(expectedSettings, editorSettings.Current);
+    }
+
+    [Fact]
+    public void ClientSettingsChangedTriggersChangedEvent()
+    {
+        // Arrange
+        var settingsManagerMock = new Mock<IClientSettingsManager>(MockBehavior.Strict);
+        settingsManagerMock
+            .Setup(x => x.GetClientSettings())
+            .Returns(ClientSettings.Default);
+
+        var editorSettings = new WorkspaceEditorSettings(settingsManagerMock.Object);
+        var settingsAccessor = new TestAccessor(editorSettings);
+
+        // Act
+        var called = false;
+        editorSettings.Changed += (caller, args) => called = true;
+
+        settingsManagerMock.Raise(
+            x => x.ClientSettingsChanged += null,
+            new ClientSettingsChangedEventArgs(ClientSettings.Default));
+
+        // Assert
+        Assert.True(called);
+    }
+
+    [Fact]
+    public void ClientSettingsChangedIsOnlyHookedOnceForMultipleListeners()
+    {
+        // Arrange
+        var settingsManagerMock = new Mock<IClientSettingsManager>(MockBehavior.Strict);
+        settingsManagerMock
+            .SetupAdd(x => x.ClientSettingsChanged += delegate { });
+        settingsManagerMock
+            .SetupRemove(x => x.ClientSettingsChanged -= delegate { });
+
+        var editorSettings = new WorkspaceEditorSettings(settingsManagerMock.Object);
+        var accessor = new TestAccessor(editorSettings);
+
+        static void Listener1(object caller, ClientSettingsChangedEventArgs args)
+        {
+        }
+
+        static void Listener2(object caller, ClientSettingsChangedEventArgs args)
+        {
+        }
+
+        // Act 1
+        editorSettings.Changed += Listener1;
+        editorSettings.Changed += Listener2;
+
+        // Assert 1 - There should be two listeners and ClientSettingsChanged should be attached.
+        Assert.Equal(2, accessor._listenerCount);
+
+        settingsManagerMock.VerifyAdd(
+            x => x.ClientSettingsChanged += It.IsAny<EventHandler<ClientSettingsChangedEventArgs>>(),
+            Times.Once());
+
+        // Act 2
+        editorSettings.Changed -= Listener1;
+        editorSettings.Changed -= Listener2;
+
+        // Assert 2 - There should be no more listeners and ClientSettingsChanged should be detached
+        Assert.Equal(0, accessor._listenerCount);
+
+        settingsManagerMock.VerifyRemove(
+            x => x.ClientSettingsChanged -= It.IsAny<EventHandler<ClientSettingsChangedEventArgs>>(),
+            Times.Once());
+    }
+
+    private class TestAccessor(WorkspaceEditorSettings instance) : TestAccessor<WorkspaceEditorSettings>(instance)
+    {
+        public int _listenerCount
+        {
+            get => Dynamic._listenerCount;
```

Comment:
Of course not, but it's a unit test. You'll get a failure if something changes, and unit tests should probably be run before submitting a PR anyway. FWIW, the philosophy of where a test accessor should live is different between the .NET runtime/libraries and Roslyn. I'm generally OK with both. What I _really_ dislike is when the production code has been contorted with lots of "for tests" comments. Or, when I see fields made internal just so a unit test can muck with them. To me, that crosses the line of writing poor product code that breaks proper type design principles to make something "testable". IMO, we shouldn't ever have any "internal for tests" comments. Instead, a test accessor should be written. That can be done in product code with more work, of it can be done using dynamic dispatch in unit tests. After all, our unit tests already do all sorts of things that'll fail without compiler complaint through Moq.

Reply:
> What I really dislike is when the production code has been contorted with lots of "for tests" comments. Or, when I see fields made internal just so a unit test can muck with them.

100% vehement agreement